### DISCUSSION TITLE
29 OCT DEV SESSION

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1451] [General]
When: 29 OCT--- 2000 UTC+8/1300 CET/1200 GMT/0800 ET

Where: [via Matrix](https://matrix.to/#/#mesa-dev:matrix.org)

Proposed Agenda:

- #1406 (Where to store ABM videos) Results ([Please vote](https://github.com/projectmesa/mesa/discussions/1411#discussioncomment-3780365)) 
- NUMFOCUS Small Business Grant status
-  DataCollection for multiple agent types (#1419/#1142)
- "Visualizing agent-level variables" (#1145)
- a roadmap to the next minor release (1.2.0)
- Mesa- Geo 
- Tutorial Status

@jackiekazil @rht @Corvince @wang-boyu @Tortar @aiwhoo @nunezco2 @EwoutH  Please feel free to add or takeaway from this agenda

**Comment by EwoutH** [comment]
Thanks for the invite, noted, and if nothing major comes in between I will join!

I will try to take some time to think about agenda items.

Also on November 14th a new education period starts at the TU Delft, including a ABM masters course. So if a new release would be planned around that time, that might be a nice target date.

_PS:_ Small nitpick, your message now says 29 September in the body.

**Reply by tpike3** [reply]
Great and fixed, thanks!

**Reply by EwoutH** [reply]
I would like to add to the agenda:
 - a roadmap to the next minor release (1.2.0)
 - DataCollection for multiple agent types ([#1419](https://github.com/projectmesa/mesa/issues/1419) / [#1142](https://github.com/projectmesa/mesa/pull/1142))

**Reply by tpike3** [reply]
Great ideas (particularly building a roadmap) and added

**Reply by EwoutH** [reply]
I would also like to add something along the line of "Visualizing agent-level variables". Currently we only support histograms for plotting agent level variables. Maybe #1145 is too complex, but something should be possible here.

**Reply by aiwhoo** [reply]
> I would also like to add something along the line of "Visualizing agent-level variables". Currently we only support histograms for plotting agent level variables. Maybe #1145 is too complex, but something should be possible here.

Clicking on an agent and having diagnostics on its Values would be amazing

**Reply by EwoutH** [reply]
I was thinking about graphs and such, but that's an amazing idea. NetLogo has this:

<img width="324" alt="Screenshot_794" src="https://user-images.githubusercontent.com/15776622/196967264-7b48e74f-f9c6-42de-933c-2dbb6001be80.png"><img width="361" alt="Screenshot_795" src="https://user-images.githubusercontent.com/15776622/196967269-4cc897e9-1569-4d6d-93ad-e221d886a91b.png">

I've no idea if it's doable with a web interface.

Maybe we should create a separate issue to discuss this idea?

**Reply by EwoutH** [reply]
Yesterday I played with some data to try to visualize the distribution of agent variable in the population over time. Without the time axis, you could just create a histogram or box plot, but over time that’s different. 

For a model with 200 agents ran 100 steps, I came up with this, which gives quite a lot of insight I think:

![45EE8CC3-D91B-44F6-ABAE-9BB40C9835CD](https://user-images.githubusercontent.com/15776622/197147143-b30f84d0-24f5-415a-88c2-971c15c13861.png)

All steps to generate this chart are written up here: https://stackoverflow.com/a/74145938/19370072

I’m quite sure this [can be made](https://github.com/chartjs/Chart.js/discussions/10368) with Chart.js, but it will be a bit of an engineering effort.

**Reply by rht** [reply]
> I've no idea if it's doable with a web interface.
> Maybe we should create a separate issue to discuss this idea?

It's doable, and yeah, should be in a separate issue.

That time series of the distribution definitely illuminates the info better than a 3D contour plot (which packs more information, but doesn't display the outline of the evolution of the percentiles). In general, there are definitely lots of information visualization routines for ABM that can be canned to a library, but haven't been done yet. It's like being able to visualize various properties of a DataFrame in much fewer lines of code with Seaborn than Matplotlib.

**Reply by rht** [reply]
We could switch to Plotly.js or D3.js instead of Chart.js.

**Reply by Corvince** [reply]
Nice chart @EwoutH !

Regarding migrating away from charts.js: I would not use d3.js, I think it is too low level. Plotly could be nice though

**Reply by EwoutH** [reply]
Side note: I just discovered such plots with error bands are extremely easy with seaborn: https://github.com/mwaskom/seaborn/issues/3101#issuecomment-1286873815

**Reply by rht** [reply]
One advantage of using D3.js is that it is already vendored in Mesa. There is also [Observable Plot](https://github.com/observablehq/plot), built by D3.js team.

Alternatively, we could create the plot within Python be it via Matplotlib or Seaborn, as is already made possible with #1139.

**Reply by tpike3** [reply]
> > I would also like to add something along the line of "Visualizing agent-level variables". Currently we only support histograms for plotting agent level variables. Maybe #1145 is too complex, but something should be possible here.
> 
> Clicking on an agent and having diagnostics on its Values would be amazing

Added

**Reply by rht** [reply]
I had to search for "seaborn" in order to find this thread. I think we should make a new discussion thread for this.

**Comment by wang-boyu** [comment]
Would also need to confirm what's next for https://github.com/projectmesa/mesa/discussions/1406?

It was listed in item 1 in [last month's session notes](https://github.com/projectmesa/mesa/discussions/1411#discussioncomment-3777868). Regarding the vote, it seems that YouTube playlist is slightly more favorable.

**Reply by EwoutH** [reply]
Thanks for bringing this up, I just voted and I would argue _strongly_ against storing them in the repo. It will increase the file size significantly, bloating everything from local clones to 

Youtube or Vimeo is perfectly fine. It offloads future transcoding. We can embed them in a [GitHub Pages](https://pages.github.com/) site, link to them in Markdown files and/or embed them [in Read the Docs](https://github.com/readthedocs/readthedocs.org/issues/879#issuecomment-51875315).

**Reply by tpike3** [reply]
Added... Thanks!

**Reply by rht** [reply]
> It will increase the file size significantly, bloating everything from local clones

It will be a separate Git repo just for the videos. And the commit history can be pruned to be at most 20-50 commits.

**Reply by EwoutH** [reply]
As some sort of archive, why not. But I would mainly serve them from YouTube.

**Reply by rht** [reply]
@wang-boyu said uploading to YouTube can't be automated. Ideally, we would gradually automate the video creation and update process.

**Reply by rht** [reply]
Those videos are usually very short. YouTube should be reserved for longer tutorials with naration.

**Comment by EwoutH** [comment]
Agenda point for next time: Go through list of open Issues and PRs. Just to check we're not missing anything.

**Comment by jackiekazil** [comment]
We used to do this. I wonder if we should do this prior to the meeting
asynchronous and then leave 10-15 minutes for anyone to advocate for a
particular one. Thoughts?

**Reply by EwoutH** [reply]
Since we're only meeting once a month I would suggest going though the list during the meeting. Every issue deserves a moment of attention at least once a month, otherwise we should close it (as "won't fix or otherwise). Idem dito for PRs.

For items requiring more time/thought it's nice to put them on the agenda beforehand, so everyone can read into it and think a bit about it.

**Comment by tpike3** [comment]
### Dev Session Notes

1. **Where to store the Mesa videos?**

- Based on discussion and related votes the decision was store videos on youtube in a Mesa channel possibly with playlists. To this end I have set up a [Project Mesa youtube channel](https://www.youtube.com/channel/UCI4lPvyQe8Xi3OlaK88vzYw) with associated projectmesaABMs@protonomail.com (projectmesa was taken) email.

2. **Mesa Examples** 
- Discussion centered on accessibility to new users so they can easily run examples. To this end, it seemed to be the best way forward was to put current examples in the example repo and make more user friendy.  **TODO:** Move examples to examples repo and make more user friendly.     

3. **NumFOCUS Grant**

- We did not get selected for a grant this go around, but got some feedback and will update our proposal and submit for next round. I am cautiously optimistic we are vectoring in to success.   

4. **DataCollection and Visualization for Activation by Type**

- We let #1145 die by neglect, which is not ok, considering we are an open source volunteer based group.  To this end, as noted in the [dev session discussion](https://github.com/projectmesa/mesa/discussions/1451#discussioncomment-4006176). We should monthly go over pull request and issues. So, we will probably need to spend the next dev session just cleaning up. **TODO:** Prior to next dev (3 DEC) sessions clean up issues and PRs to reduce the number.   
- Adding datacollection by agent as a feature, the discussion was taking too long so we want to hold a separate call to discuss agent collection by attribute and associated visualization. #1419 and #1145. **TODO:** Schedule meet up challenge is timing, so how about, Thursday the 10th of NOV, at 1900 UTC+8/1200 CET/1100 GMT/0600 ET?

5. **Roadmap to Next Release**

- Make sure we continue to use milestones so we can track actions, this will also help ensure we do not drop the ball and neglect PRs or issues. To this end I have updated Milestone Release and major efforts. **TODO:** Please update and our change milestones as you see fit

6. **Mesa-Geo**

- Got its first paper accepted (citation)[https://github.com/projectmesa/mesa-geo/blob/main/CITATION.bib]!!!; And even has a tentative cite from a Mesa paper now under draft for the Autonomous Agents and Multiagent Systems conference 

7.   **Tutorial**

- First three coding videos submitted to SFI (session 3-6), waiting for feedback and then will proceed, script is mostly done with some updates enroute. [Can be seen in Google COLAB, here is the repo](https://github.com/edgeofchaos42/ComplexityExplorer)


Feel free to make comments, critique, add to as you see fit. Happy to make adjustments if a mischaracterized or did not accurately capture something.

**Comment by rht** [comment]
The "s" ending in projectmesaABMs@protonmail.com is going to trip some people up.

**Reply by tpike3** [reply]
@rht  dang, your right... well we can always update it but really it is just a dummy email to open the youtube account. I don't think it will get too much traffic.
